File: baekjoon/1753.py
import sys
from queue import PriorityQueue

input = sys.stdin.readline

v, e = map(int, input().split())
start = int(input())

graph = [[] for _ in range(v)]  # |시작노드, 가중치, 도착노드|

visited = [float("inf") for _ in range(v)]


def dijkstra(start) -> None:
    pq = PriorityQueue()
    pq.put((0, start))  # |거리 , 정점|
    visited[start] = 0

    while not pq.empty():
        dist, curr = pq.get()

        if visited[curr] < dist:
            continue

        for weight, next_ in graph[curr]:
            next_dist = dist + weight
            if next_dist < visited[next_]:
                visited[next_] = next_dist
                pq.put((next_dist, next_))


for _ in range(e):
    x, y, z = map(int, input().split())

    graph[x - 1].append([z, y - 1])

dijkstra(start - 1)


for i in visited:
    if i == float("inf"):
        print("INF")
    else:
        print(i)

File: baekjoon/1753.cpp
#include <cstdio>
#include <queue>
#include <vector>
#include <functional>

using namespace std;

void dijkstra(int num, const vector<vector<pair<int, int>>> &graph)
{
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    vector<int> distances(graph.size(), 1e9);
    distances[num] = 0;
    pq.push({0, num});

    while (!pq.empty())
    {
        int dist = pq.top().first;
        int current = pq.top().second;
        pq.pop();

        if (dist > distances[current])
            continue;

        for (const auto &edge : graph[current])
        {
            int next = edge.first;
            int weight = edge.second;

            if (dist + weight < distances[next])
            {
                distances[next] = dist + weight;
                pq.push({distances[next], next});
            }
        }
    }

    for (int i = 1; i < distances.size(); i++)
    {
        if (distances[i] == 1e9)
            printf("INF\n");
        else
            printf("%d\n", distances[i]);
    }
}

int main()
{
    int v, e, k;
    scanf("%d %d", &v, &e);
    scanf("%d", &k);
    vector<vector<pair<int, int>>> graph(v + 1);

    for (int i = 0; i < e; i++)
    {
        int x, y, z;
        scanf("%d %d %d", &x, &y, &z);
        graph[x].emplace_back(y, z);
    }
    dijkstra(k, graph);
}